File: ios/Zest/Services/UnitConverter.swift
import Foundation

class UnitConverter {
    static let shared = UnitConverter()
    
    private let conversions: [String: [String: Double]] = [
        "g": ["g": 1, "kg": 0.001, "oz": 0.035274, "lb": 0.00220462],
        "kg": ["g": 1000, "kg": 1, "oz": 35.274, "lb": 2.20462],
        "oz": ["g": 28.3495, "kg": 0.0283495, "oz": 1, "lb": 0.0625],
        "lb": ["g": 453.592, "kg": 0.453592, "oz": 16, "lb": 1],
        "ml": ["ml": 1, "l": 0.001, "cup": 0.00422675, "tbsp": 0.067628, "tsp": 0.202884],
        "l": ["ml": 1000, "l": 1, "cup": 4.22675, "tbsp": 67.628, "tsp": 202.884],
        "cup": ["ml": 236.588, "l": 0.236588, "cup": 1, "tbsp": 16, "tsp": 48],
        "tbsp": ["ml": 14.7868, "l": 0.0147868, "cup": 0.0625, "tbsp": 1, "tsp": 3],
        "tsp": ["ml": 4.92892, "l": 0.00492892, "cup": 0.0208333, "tbsp": 0.333333, "tsp": 1]
    ]
    
    func convert(amount: Double, from fromUnit: String, to toUnit: String) -> Double {
        if fromUnit == toUnit { return amount }
        guard let conversion = conversions[fromUnit]?[toUnit] else { return amount }
        return round(amount * conversion * 100) / 100
    }
    
    func scaleRecipe(amount: Double, originalServings: Int, newServings: Int) -> Double {
        let scale = Double(newServings) / Double(originalServings)
        return round(amount * scale * 100) / 100
    }
}

File: android/app/src/main/java/com/neostudios/zest/domain/usecase/UnitConverter.kt
package com.neostudios.zest.domain.usecase

object UnitConverter {
    private val conversions = mapOf(
        "g" to mapOf("g" to 1.0, "kg" to 0.001, "oz" to 0.035274, "lb" to 0.00220462),
        "kg" to mapOf("g" to 1000.0, "kg" to 1.0, "oz" to 35.274, "lb" to 2.20462),
        "oz" to mapOf("g" to 28.3495, "kg" to 0.0283495, "oz" to 1.0, "lb" to 0.0625),
        "lb" to mapOf("g" to 453.592, "kg" to 0.453592, "oz" to 16.0, "lb" to 1.0),
        "ml" to mapOf("ml" to 1.0, "l" to 0.001, "cup" to 0.00422675, "tbsp" to 0.067628, "tsp" to 0.202884),
        "l" to mapOf("ml" to 1000.0, "l" to 1.0, "cup" to 4.22675, "tbsp" to 67.628, "tsp" to 202.884),
        "cup" to mapOf("ml" to 236.588, "l" to 0.236588, "cup" to 1.0, "tbsp" to 16.0, "tsp" to 48.0),
        "tbsp" to mapOf("ml" to 14.7868, "l" to 0.0147868, "cup" to 0.0625, "tbsp" to 1.0, "tsp" to 3.0),
        "tsp" to mapOf("ml" to 4.92892, "l" to 0.00492892, "cup" to 0.0208333, "tbsp" to 0.333333, "tsp" to 1.0)
    )
    
    fun convert(amount: Double, fromUnit: String, toUnit: String): Double {
        if (fromUnit == toUnit) return amount
        val conversion = conversions[fromUnit]?.get(toUnit) ?: return amount
        return (amount * conversion * 100).toInt() / 100.0
    }
    
    fun scaleRecipe(amount: Double, originalServings: Int, newServings: Int): Double {
        val scale = newServings.toDouble() / originalServings
        return (amount * scale * 100).toInt() / 100.0
    }
}
